File: packages/sdk-python/agentlens/span.py
"""Span — mutable object passed to trace callbacks and context managers."""

from __future__ import annotations

import datetime
from typing import Any, Optional, Union

from .models import SpanData


class Span:
    """A mutable span annotated by the caller during an LLM or agent step.

    Mirrors the TypeScript ``Span`` class API surface.  Fluent setters
    return ``self`` so calls can be chained::

        span.set_model("gpt-4o", "openai").set_tokens(100, 50)

    The span is closed automatically by :class:`~agentlens.tracer.Tracer`
    (or the context manager returned by :meth:`AgentLens.trace`) once the
    body finishes or raises.
    """

    def __init__(
        self,
        span_id: str,
        trace_id: str,
        name: str,
        project_id: str,
        parent_span_id: Optional[str] = None,
    ) -> None:
        self._span_id = span_id
        self._trace_id = trace_id
        self._name = name
        self._project_id = project_id
        self._parent_span_id = parent_span_id
        self._started_at: datetime.datetime = datetime.datetime.now(datetime.timezone.utc)

        self._model: Optional[str] = None
        self._provider: Optional[str] = None
        self._input: Optional[str] = None
        self._output: Optional[str] = None
        self._input_tokens: Optional[int] = None
        self._output_tokens: Optional[int] = None
        self._cost_usd: Optional[float] = None
        self._latency_ms: Optional[int] = None
        self._status: str = "success"
        self._error_message: Optional[str] = None
        self._metadata: dict[str, Any] = {}
        self._ended_at: Optional[datetime.datetime] = None

    # ── Fluent setters ──────────────────────────────────────────────────────────

    def set_input(self, data: str) -> "Span":
        """Set the raw LLM prompt or tool input (stored in Elasticsearch only)."""
        self._input = data
        return self

    def set_output(self, data: str) -> "Span":
        """Set the raw LLM completion or tool output (stored in Elasticsearch only)."""
        self._output = data
        return self

    def set_metadata(self, key: str, value: Any) -> "Span":
        """Attach a free-form metadata key-value pair to this span."""
        self._metadata[key] = value
        return self

    def set_model(self, model: str, provider: Optional[str] = None) -> "Span":
        """Set the model name and optional provider (e.g. ``'gpt-4o'``, ``'openai'``)."""
        self._model = model
        if provider is not None:
            self._provider = provider
        return self

    def set_tokens(
        self,
        input_tokens: int,
        output_tokens: int,
        cost_usd: Optional[float] = None,
    ) -> "Span":
        """Set token counts and optionally the USD cost."""
        self._input_tokens = input_tokens
        self._output_tokens = output_tokens
        if cost_usd is not None:
            self._cost_usd = cost_usd
        return self

    def set_error(self, error: Union[Exception, str]) -> "Span":
        """Record an error on this span; automatically sets status to ``'error'``."""
        self._status = "error"
        self._error_message = str(error) if isinstance(error, Exception) else error
        return self

    # ── Lifecycle ───────────────────────────────────────────────────────────────

    def end(self) -> None:
        """Close the span: record ``ended_at`` and compute ``latency_ms``.

        Idempotent — subsequent calls are no-ops.
        """
        if self._ended_at is not None:
            return
        self._ended_at = datetime.datetime.now(datetime.timezone.utc)
        delta = self._ended_at - self._started_at
        self._latency_ms = int(delta.total_seconds() * 1000)

    # ── Serialisation ───────────────────────────────────────────────────────────

    def to_dict(self) -> dict[str, Any]:
        """Serialise to the camelCase wire format expected by the ingest endpoint.

        :meth:`end` should be called before :meth:`to_dict`.
        """
        return SpanData(
            span_id=self._span_id,
            trace_id=self._trace_id,
            project_id=self._project_id,
            name=self._name,
            status=self._status,  # type: ignore[arg-type]
            metadata=dict(self._metadata),
            started_at=self._started_at.isoformat(),
            parent_span_id=self._parent_span_id,
            model=self._model,
            provider=self._provider,
            input=self._input,
            output=self._output,
            input_tokens=self._input_tokens,
            output_tokens=self._output_tokens,
            cost_usd=self._cost_usd,
            latency_ms=self._latency_ms,
            error_message=self._error_message,
            ended_at=self._ended_at.isoformat() if self._ended_at else None,
        ).to_wire()

File: packages/sdk-core/src/span.ts
import type { SpanData, SpanStatus } from './types.js';

/**
 * A mutable span object passed into the `trace()` callback.
 * Callers use the fluent setters to annotate LLM calls; the tracer calls
 * `end()` automatically after the callback resolves or rejects.
 */
export class Span {
  private readonly _spanId: string;
  private readonly _traceId: string;
  private readonly _parentSpanId: string | undefined;
  private readonly _projectId: string;
  private readonly _name: string;
  private readonly _startedAt: Date;

  private _model?: string;
  private _provider?: string;
  private _input?: string;
  private _output?: string;
  private _inputTokens?: number;
  private _outputTokens?: number;
  private _costUsd?: number;
  private _latencyMs?: number;
  private _status: SpanStatus = 'success';
  private _errorMessage?: string;
  private _metadata: Record<string, unknown> = {};
  private _endedAt?: Date;

  constructor(
    spanId: string,
    traceId: string,
    name: string,
    projectId: string,
    parentSpanId?: string,
  ) {
    this._spanId = spanId;
    this._traceId = traceId;
    this._name = name;
    this._projectId = projectId;
    this._parentSpanId = parentSpanId;
    this._startedAt = new Date();
  }

  // ── Fluent setters ──────────────────────────────────────────────────────────

  /**
   * Sets the raw LLM prompt or tool input.
   * Stored in Elasticsearch only — not persisted to PostgreSQL.
   */
  setInput(data: string): this {
    this._input = data;
    return this;
  }

  /**
   * Sets the raw LLM completion or tool output.
   * Stored in Elasticsearch only — not persisted to PostgreSQL.
   */
  setOutput(data: string): this {
    this._output = data;
    return this;
  }

  /** Attaches a free-form metadata key-value pair to this span. */
  setMetadata(key: string, value: unknown): this {
    this._metadata[key] = value;
    return this;
  }

  /** Sets model and provider (e.g. `'gpt-4o'`, `'openai'`). */
  setModel(model: string, provider?: string): this {
    this._model = model;
    if (provider !== undefined) this._provider = provider;
    return this;
  }

  /** Sets token counts and optionally the USD cost. */
  setTokens(inputTokens: number, outputTokens: number, costUsd?: number): this {
    this._inputTokens = inputTokens;
    this._outputTokens = outputTokens;
    if (costUsd !== undefined) this._costUsd = costUsd;
    return this;
  }

  /**
   * Records an error on this span.
   * Automatically sets `status` to `'error'` and captures the message.
   */
  setError(error: Error | string): this {
    this._status = 'error';
    this._errorMessage = error instanceof Error ? error.message : error;
    return this;
  }

  // ── Lifecycle ───────────────────────────────────────────────────────────────

  /**
   * Closes the span: records `endedAt` and computes `latencyMs`.
   * Called automatically by the tracer after the callback resolves/rejects.
   * Calling it manually is safe — subsequent calls are no-ops.
   */
  end(): void {
    if (this._endedAt) return;
    this._endedAt = new Date();
    this._latencyMs = this._endedAt.getTime() - this._startedAt.getTime();
  }

  // ── Serialization ───────────────────────────────────────────────────────────

  /**
   * Serializes this span to the wire format expected by the ingest endpoint.
   * `end()` should be called before `toJSON()`.
   */
  toJSON(): SpanData {
    const data: SpanData = {
      spanId: this._spanId,
      traceId: this._traceId,
      projectId: this._projectId,
      name: this._name,
      status: this._status,
      metadata: { ...this._metadata },
      startedAt: this._startedAt.toISOString(),
    };

    if (this._parentSpanId !== undefined) data.parentSpanId = this._parentSpanId;
    if (this._model !== undefined) data.model = this._model;
    if (this._provider !== undefined) data.provider = this._provider;
    if (this._input !== undefined) data.input = this._input;
    if (this._output !== undefined) data.output = this._output;
    if (this._inputTokens !== undefined) data.inputTokens = this._inputTokens;
    if (this._outputTokens !== undefined) data.outputTokens = this._outputTokens;
    if (this._costUsd !== undefined) data.costUsd = this._costUsd;
    if (this._latencyMs !== undefined) data.latencyMs = this._latencyMs;
    if (this._errorMessage !== undefined) data.errorMessage = this._errorMessage;
    if (this._endedAt !== undefined) data.endedAt = this._endedAt.toISOString();

    return data;
  }
}
